contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,198 @@
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 internal _latestVersion;
+
+    // Mappings that keep track of the protocol version and corresponding contract address.
+    mapping(uint256 => address) internal _versionToAddress;
+    mapping(address => uint256) internal _addressToVersion;
+
+    /**
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    /**
+     * @dev Initializes the contract by setting `_blockchainID` and `_latestVersion`.
+     * Also adds the initial protocol versions to the registry.
+     */
+    constructor(ProtocolRegistryEntry[] memory initialEntries) {
+        _blockchainID = WARP_MESSENGER.getBlockchainID();
+        _latestVersion = 0;
+
+        for (uint256 i = 0; i < initialEntries.length; i++) {
+            _addToRegistry(initialEntries[i]);
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a warp out-of-band message, and adds the new protocol version
+     * address to the registry.
+     * If a version is greater than the current latest version, it will be set as the latest version.
+     * If a version is less than the current latest version, it is added to the registry, but
+     * doesn't change the latest version.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Requirements:
+     *
+     * - a valid Warp out-of-band message must be provided.
+     * - Message source chain ID must be the same as the blockchain ID of the registry.
+     * - Message origin sender address must be the same as the `VALIDATORS_SOURCE_ADDRESS`.
+     * - Message destination chain ID must be the same as the blockchain ID of the registry.
+     * - Message destination address must be the same as the address of the registry.
+     * - the version must not be zero.
+     * - the version must not already be registered.
+     * - protocol address must not be zero address.
+     */
+    function addProtocolVersion(uint32 messageIndex) external virtual {
+        // Get and validate for a warp out-of-band message.
+        (WarpMessage memory message, bool success) = WARP_MESSENGER
+            .getVerifiedWarpMessage(messageIndex);
+        require(success, "WarpProtocolRegistry: invalid warp message");
+        require(
+            message.sourceChainID == _blockchainID,
+            "WarpProtocolRegistry: invalid source chain ID"
+        );
+        // Check that the message is sent through a warp out of band message.
+        require(
+            message.originSenderAddress == VALIDATORS_SOURCE_ADDRESS,
+            "WarpProtocolRegistry: invalid origin sender address"
+        );
+        require(
+            message.destinationChainID == _blockchainID,
+            "WarpProtocolRegistry: invalid destination chain ID"
+        );
+        require(
```

Comment:
I'm think of a future event when we want to roll out a TeleporterMessenger upgrade. It would be really nice if the registry contract was deployed to the same contract address on most (all) chains such that the out-of-band warp message needed to be signed by the validators of each chain was the same.

We could deploy the registry using Nick's method, but to make that less error prone, similar to the `TeleporterMessenger` contract itself, we should move external calls out of the constructor to make it less likely for the deployment to fail. Thoughts on always calling `warp.getBlockchainID` here rather than storing the blockchain ID in the contract?

Reply:
Added issue and extra considerations in new issue: https://github.com/ava-labs/teleporter/issues/76